ddl/placement_sql_test.go

```diff
@@ -523,3 +523,121 @@ add placement policy
+add placement policy
+	constraints='["+zone=sh"]'
+	role=leader
+	replicas=1`)
+	c.Assert(err, IsNil)
+
+	bundles := make(map[string]*placement.Bundle)
+	is := s.dom.InfoSchema()
+	is.MockBundles(bundles)
+
+	tb, err := is.TableByName(model.NewCIStr("test"), model.NewCIStr("t1"))
+	c.Assert(err, IsNil)
+	partDefs := tb.Meta().GetPartitionInfo().Definitions
+
+	for _, def := range partDefs {
+		if def.Name.String() == "p0" {
+			groupID := placement.GroupID(def.ID)
+			bundles[groupID] = &placement.Bundle{
+				ID: groupID,
+				Rules: []*placement.Rule{
+					{
+						GroupID: groupID,
+						Role:    placement.Leader,
+						Count:   1,
+						LabelConstraints: []placement.LabelConstraint{
+							{
+								Key:    placement.DCLabelKey,
+								Op:     placement.In,
+								Values: []string{"bj"},
+							},
+						},
+					},
+				},
+			}
+			break
+		}
+	}
+	dbInfo := testGetSchemaByName(c, tk.Se, "test")
+	tk2 := testkit.NewTestKit(c, s.store)
+	var chkErr error
+	done := false
+	testcases := []struct {
+		name      string
+		hook      *ddl.TestDDLCallback
+		expectErr error
+	}{
+		{
+			name: "write partition p0 during StateGlobalTxnOnly",
+			hook: func() *ddl.TestDDLCallback {
+				hook := &ddl.TestDDLCallback{}
+				hook.OnJobUpdatedExported = func(job *model.Job) {
+					if job.Type == model.ActionAlterTableAlterPartition && job.State == model.JobStateRunning &&
+						job.SchemaState == model.StateGlobalTxnOnly && job.SchemaID == dbInfo.ID && done == false {
+						done = true
+						tk2.MustExec("use test")
+						tk2.MustExec("set @@txn_scope=bj")
+						_, chkErr = tk2.Exec("insert into t1 (c) values (1);")
+					}
+				}
+				return hook
+			}(),
+			expectErr: fmt.Errorf(".*under StateGlobalTxnOnly.*"),
+		},
+		// FIXME: support abort read txn during StateGlobalTxnOnly
+		//{
+		//	name: "read partition p0 during middle state",
+		//	hook: func() *ddl.TestDDLCallback {
+		//		hook := &ddl.TestDDLCallback{}
+		//		hook.OnJobUpdatedExported = func(job *model.Job) {
+		//			if job.Type == model.ActionAlterTableAlterPartition && job.State == model.JobStateRunning &&
+		//				job.SchemaState == model.StateGlobalTxnOnly && job.SchemaID == dbInfo.ID && done == false {
+		//				done = true
+		//				tk2.MustExec("use test")
+		//				tk2.MustExec("set @@txn_scope=bj")
+		//				tk2.MustExec("begin;")
+		//				tk2.MustExec("select * from t1 where c < 6;")
+		//				_, chkErr = tk2.Exec("commit")
+		//			}
+		//		}
+		//		return hook
+		//	}(),
+		//	expectErr: fmt.Errorf(".*under StateGlobalTxnOnly.*"),
+		//},
+	}
+	originalHook := s.dom.DDL().GetHook()
+	for _, testcase := range testcases {
+		c.Log(testcase.name)
+		done = false
+		s.dom.DDL().(ddl.DDLForTest).SetHook(testcase.hook)
+		_, err = tk.Exec(`alter table t1 alter partition p0
+alter placement policy
+	constraints='["+zone=bj"]'
+	role=leader
+	replicas=1`)
+		c.Assert(err, IsNil)
+		s.dom.DDL().(ddl.DDLForTest).SetHook(originalHook)
+		c.Assert(done, Equals, true)
```

Comment:
I don't understand. How did it modify the `done` in this loop?

Reply:
In each loop, the hook would be passed by the `testcase`, and the `done` is set to `true` in each hook.